[#5759] Prevent queries in replica creation/finalize (4-2-stable)

alanking opened the pull request:
Tried to split up the commits to make it a bit more manageable but it's still a big change. The bulk of it is moving the contents of `rsPhyPathReg` into the new API plugin and adding some documentation, so there are only a few critical areas to look at.

CI tests passed, federation passed, topology-from-provider passed, topology-from-consumer running

trel commented on scripts/irods/test/test_iput.py:
```diff
@@ -513,6 +513,48 @@ def do_test_issue_5496_impl(self, file_size_in_bytes):
+            for name in data_names:
+                logical_path = os.path.join(collection_path, name)
+
+                self.user.assert_icommand(['iput', local_file, logical_path])
```
does `ireg` need the same exercising?

alanking replied:
I started to write tests for several iCommands (`irepl`, `icp`, etc.) but backed them out since I was just testing the apostrophe bit (didn't seem important to me at the time to try all the different iCommands). I can add these back as well as one for `ireg`.

alanking replied:
I also noticed while contemplating this test that this does not cover all the permutations for `icp` in particular because there will be two data objects to consider, not just one: source and destination. So that test will need to cover an additional set of scenarios for the source data object.

alanking replied:
Done

trel commented on server/api/include/rs_data_object_finalize.hpp:
```diff
@@ -13,12 +13,34 @@ extern "C" {
-/// is required in the form of a JSON structure.
+/// \parblock
+/// The replicas field is a complete set of the columns in their current state as well as the
+/// desired modifications is required in the form of a JSON structure. Each replica can also
```
doesn't quite parse as english.

alanking replied:
Hmmm... I just copied and pasted this from `lib/api/include/data_object_finalize.hpp`. Not sure how that happened. Fixed.

trel commented on server/api/src/rsDataObjOpen.cpp:
```diff
@@ -327,27 +338,94 @@ namespace
                         irods::log(LOG_ERROR, fmt::format("[{}:{}] - failed to unlock data object [error_code={}]", __FUNCTION__, __LINE__, unlock_ec));
                     }
+
+                    if (rst::contains(_existing_replica_list->dataId)) {
```
can we get a comment here?

alanking replied:
Done

trel commented on server/api/src/rs_register_physical_path.cpp:
```diff
@@ -0,0 +1,32 @@
+    out = nullptr;
+
+    return rc;
+} // rs_data_object_finalize
```
wrong name

alanking replied:
Done

trel commented on server/core/include/irods_file_object.hpp:
```diff
@@ -197,6 +199,22 @@ namespace irods {
+    /// data_object_proxy_t has been generated.
+    /// \endparblock
+    ///
+    /// \param[in] _comm
```
moar

alanking replied:
Done

trel commented on server/core/src/physPath.cpp:
```diff
@@ -429,7 +429,7 @@ getchkPathPerm( rsComm_t *rsComm, dataObjInp_t *dataObjInp,
 
 #if 1
-    // Possibly need to revisit this -- this is wrong
+    // TODO: this is wrong
```
why?   create / link to an issue?

alanking replied:
Created #5775 for this

trel commented on unit_tests/src/test_data_object_finalize.cpp:
```diff
@@ -111,6 +111,8 @@ TEST_CASE("finalize", "[finalize]")
             // Get data object info, modify some fields in each replica, and stamp the catalog
             json input;
 
+            input["logical_path"] = og_op.logical_path();
```
og_op, pretty funny.

alanking replied:
I do what I can

trel reviewed (commented):
easier to follow than i expected when opening this tab.     good stuff.

alanking commented:
Addressed review comments except for adding the extra tests. I addressed all of the Codacy issues except for:

1. The docstring issue (this is a configuration issue in Codacy that needs to be resolved)
2. The `irods::at_scope_exit` being unused issue, which is just the nature of the object

Next push will contain new tests and then I'll report test results

alanking commented:
Added tests for `icp`, `ireg`, `irepl`, `iphymv`, and `imv`. Confirmed everything is working in topology.

korydraughn commented on server/core/src/server_utilities.cpp:
```diff
@@ -50,7 +50,7 @@ namespace irods
 
         auto* buf = static_cast<char*>(allocate(sizeof(char) * buf_size));
-        std::strncpy(buf, _s.data(), _s.length());
+        std::snprintf(buf, _s.length(), "%s", _s.data());
```
This will truncate the string being copied by one byte. Replacing `_s.length()` with `buf_size` should take care of it.

korydraughn replied:
Ignore this one. I was doing the review one commit at a time.

alanking replied:
Yep, fixed.

korydraughn commented on server/core/src/server_utilities.cpp:
```diff
@@ -39,5 +39,24 @@ namespace irods
+        const auto buf_size = _s.length() + 1;
+
+        auto* buf = static_cast<char*>(allocate(sizeof(char) * buf_size));
+        std::snprintf(buf, buf_size, "%s", _s.data());
```
What is the reasoning behind using `snprintf` over `strncpy` here?

alanking replied:
Codacy complained via the FlawFinder tool which states: "Easily used incorrectly; doesn't always \0-terminate or check for invalid pointers [MS-banned] (CWE-120)." `std::snprintf` seems to be considered a safe alternative.

korydraughn replied:
That makes sense. However, I bet this likely incurs more overhead. Maybe codacy would feel better about `std::memcpy`. In this case, we don't have to pay for the additional overhead given `buf` is cleared to zeros which gives us the null terminator for free.

korydraughn replied:
Assuming my testing is correct, I'm showing strncpy as being 4.4x faster than snprintf and memcpy being 4.3x faster than snprintf. Here is the test: https://quick-bench.com/q/axHaBt3rn3B5FIX_h5QSexwCnBM

korydraughn replied:
The only problem I see with `strncpy` is that it stops at null bytes, but given that `to_bytes_buffer` takes in a `string_view`, it is probably safe to say that is fine.

alanking replied:
Cool, I don't mind using either one, really. I will try `std::memcpy` and see how that goes.

alanking replied:
Codacy has deemed `std::memcpy` acceptable. I am okay to revert this to use `std::strncpy` as well. If we go that route, I may leave a comment in defense of its use for future readers of the Codacy issues.

korydraughn replied:
Cool. It's your call.

alanking replied:
Okay. I think I'll just revert it and leave an explanatory comment. Gracias

korydraughn reviewed (commented):
Looking good

alanking commented on plugins/api/src/data_object_finalize.cpp:
```diff
@@ -89,6 +89,19 @@ namespace
         return _api->call_handler<BytesBuf*, BytesBuf**>(_comm, _input, _output);
     } // call_data_object_finalize
 
+    auto split_before_and_after_info(const json& _json_list) -> std::tuple<json, json>
+    {
+        json before;
+        json after;
+
+        for (const auto& i : _json_list) {
+            before.push_back(i.at("before"));
+            after.push_back(i.at("after"));
+        }
+
+        return {before, after};
+    } // split_before_and_after_info
+
```
I'll look into it. I think I will have to change some interfaces and add a new factory method for `data_object_proxy` as the existing interfaces operate on `nlohmann::json::array`s.

alanking replied:
I managed to implement a version of this in which before and after are `std::vector<const nlohmann::json*>`s and it seems to be working for normal use cases. I'll run tests shortly

trel commented on scripts/irods/test/test_ichmod.py:
```diff
@@ -1,28 +1,74 @@
+        self.admin.assert_icommand(['ils', '-LA', self.collection], 'STDOUT', 'file')
+        self.user.assert_icommand(['iget', logical_path, os.path.join(self.user.local_session_dir, '')])
+
+    def test_collection_acl_inheritance_for_icp__issue_3032(self):
```
please consider `imv` into the inheritance structure as well.

alanking replied:
Done

alanking replied:
I take it back. `imv` does not work in this scenario. That's what I get for trying to go fast

alanking commented:
Squashed. Running tests again

alanking commented:
Core tests, unit tests, timing tests, and federation tests passed

trel commented:
so good.    please to # when ready.

alanking commented:
#'d

alanking commented:
Just noticed that I forgot to include the new headers in the development library. I can squash those before merging

trel commented:
yeah, let's squash it into the right place - not pressed for time on this one.
